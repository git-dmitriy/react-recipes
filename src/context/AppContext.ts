import { createContext } from 'react';
import { ContextTypes } from '@/appTypes';

export const AppContext = createContext({} as ContextTypes);
