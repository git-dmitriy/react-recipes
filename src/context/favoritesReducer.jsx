import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from './types';

const handlers = {
  [ADD_TO_FAVORITE]: (state, { payload }) => {
    return {
      favorites: [...state.favorites, payload],
    };
  },
  [REMOVE_FROM_FAVORITE]: (state, { payload }) => ({
    favorites: state.favorites.filter((item) => item.idMeal !== payload.idMeal),
  }),
  DEFAULT: (state) => state,
};

export const favoriteReducer = (state, action) => {
  const handler = handlers[action.type] || handlers.DEFAULT;
  return handler(state, action);
};
