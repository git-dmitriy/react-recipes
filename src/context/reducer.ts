import {
    ADD_TO_FAVORITE,
    REMOVE_FROM_FAVORITE,
    SWITCH_THEME,
    SET_IS_LOADING,
} from '@context/contextTypes.ts';
import {StateTypes, ActionTypes} from '@/appTypes';

export const reducer = (state: StateTypes, action: ActionTypes) => {
    let favorites = null;
    switch (action.type) {
        case ADD_TO_FAVORITE:
            favorites = [...state.favorites, action.payload];
            window.localStorage.setItem('favorites', JSON.stringify(favorites));
            return {...state, favorites};

        case REMOVE_FROM_FAVORITE:
            favorites = state.favorites.filter(
                (item) => item.idMeal !== action.payload.idMeal
            );
            window.localStorage.setItem('favorites', JSON.stringify(favorites));
            return {...state, favorites};

        case SWITCH_THEME:
            window.localStorage.setItem('theme', action.payload);
            return {
                ...state,
                theme: action.payload,
            };

        case SET_IS_LOADING:
            return {
                ...state,
                isLoading: action.payload,
            };

        default:
            return state;
    }
};
