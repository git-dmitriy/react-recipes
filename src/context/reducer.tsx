import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE, SWITCH_THEME } from './types';
import { StateTypes, ActionTypes } from 'appTypes';

export const reducer = (state: StateTypes, action: ActionTypes) => {
  let favorites = null;
  switch (action.type) {
    case ADD_TO_FAVORITE:
      favorites = [...state.favorites, action.payload];
      window.localStorage.setItem('favorites', JSON.stringify(favorites));
      return { ...state, favorites };

    case REMOVE_FROM_FAVORITE:
      favorites = state.favorites.filter(
        (item) => item.idMeal !== action.payload.idMeal
      );
      window.localStorage.setItem('favorites', JSON.stringify(favorites));
      return { ...state, favorites };

    case SWITCH_THEME:
      window.localStorage.setItem('theme', action.payload);
      return {
        ...state,
        theme: action.payload,
      };

    default:
      return state;
  }
};
