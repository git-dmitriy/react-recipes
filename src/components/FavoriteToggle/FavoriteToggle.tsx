import {
  BsFillBookmarkPlusFill,
  BsFillBookmarkCheckFill,
} from 'react-icons/bs';

import { useState, useEffect, useContext } from 'react';
import { AppContext } from 'context/AppContext';
import { MealItemTypes } from 'appTypes';

type P = {
  meal: MealItemTypes;
  isDark?: boolean;
};

export const FavoriteToggle: React.FC<P> = ({ meal, isDark = false }) => {
  const [isFavorite, setIsFavorite] = useState<boolean | null>(null);
  const { state, addToFavorites, removeFromFavorites } = useContext(AppContext);

  useEffect(
    () =>
      setIsFavorite(
        !!state.favorites.find((item) => item.idMeal === meal.idMeal)
      ),
    [state.favorites]
  );

  const addToFavorite = () => {
    addToFavorites(meal);
  };

  const removeFromFavorite = () => {
    removeFromFavorites(meal);
  };

  return (
    <div className='mx-2 mt-2'>
      {isFavorite ? (
        <button
          className='text-2xl text-red-500 transition duration-200 uppercase'
          onClick={removeFromFavorite}
        >
          <BsFillBookmarkCheckFill className='fill-current' />
        </button>
      ) : (
        <button
          className={`text-2xl text-gray-900 ${
            isDark ? '' : 'dark:text-orange-200'
          } hover:text-red-500 transition duration-200 uppercase`}
          onClick={addToFavorite}
        >
          <BsFillBookmarkPlusFill className='fill-current' />
        </button>
      )}
    </div>
  );
};
