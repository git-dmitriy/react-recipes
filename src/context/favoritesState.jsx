import React, { useReducer } from 'react';
import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from 'context/types';
import { FavoriteContext } from 'context/favoritesContext';
import { favoriteReducer } from 'context/favoritesReducer';

export const FavoritesState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
  };

  const [state, dispatch] = useReducer(favoriteReducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  return (
    <FavoriteContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        state,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
};
