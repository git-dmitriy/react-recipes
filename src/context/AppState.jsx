import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';

export const AppState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
    theme: window.localStorage.getItem('theme') || 'light',
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  return (
    <AppContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        switchTheme,
        state,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
