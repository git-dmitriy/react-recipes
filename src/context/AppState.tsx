import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
  SET_IS_LOADING,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';
import { StateTypes, MealItemTypes, ContextTypes } from 'appTypes';

type P = { children: React.ReactNode };

export const AppState = ({ children }: P) => {
  const favorites = window.localStorage.getItem('favorites') || '[]';

  const initialState: StateTypes = {
    favorites: JSON.parse(favorites),
    theme: window.localStorage.getItem('theme') || 'light',
    isLoading: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item: MealItemTypes) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item: MealItemTypes) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode: string) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  const setIsLoading = (status = false) => {
    dispatch({
      type: SET_IS_LOADING,
      payload: status,
    });
  };

  const value: ContextTypes = {
    addToFavorites,
    removeFromFavorites,
    switchTheme,
    setIsLoading,
    state,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};
