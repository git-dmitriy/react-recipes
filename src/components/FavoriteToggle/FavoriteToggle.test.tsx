import { FavoriteToggle } from './FavoriteToggle';
import renderer from 'react-test-renderer';
import { AppContext } from 'context/AppContext';
import { fireEvent, render, screen } from '@testing-library/react';

const data = {
  idMeal: '52895',
  strMeal: 'English Breakfast',
  strMealThumb:
    'https://www.themealdb.com/images/media/meals/utxryw1511721587.jpg',
};

const addToFavorites = jest.fn();
const removeFromFavorites = jest.fn();

const context = {
  state: {
    favorites: [],
    theme: 'light',
    isLoading: false,
  },
  addToFavorites,
  removeFromFavorites,
  switchTheme: jest.fn(),
  setIsLoading: jest.fn(),
};

it('should render component without errors', () => {
  const tree = renderer
    .create(
      <AppContext.Provider value={context}>
        <FavoriteToggle meal={data} />
      </AppContext.Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('should add item to favorites on click', () => {
  render(
    <AppContext.Provider value={context}>
      <FavoriteToggle meal={data} />
    </AppContext.Provider>
  );

  const favoriteToggler = screen.getByRole('button');

  fireEvent.click(favoriteToggler);

  expect(addToFavorites).toBeCalled();
});

it('should remove item from favorites on click', () => {
  const context = {
    state: {
      favorites: [
        {
          idMeal: '52895',
          strMeal: 'English Breakfast',
          strMealThumb:
            'https://www.themealdb.com/images/media/meals/utxryw1511721587.jpg',
        },
      ],
      theme: 'light',
      isLoading: false,
    },
    addToFavorites,
    removeFromFavorites,
    switchTheme: jest.fn(),
    setIsLoading: jest.fn(),
  };

  render(
    <AppContext.Provider value={context}>
      <FavoriteToggle meal={data} />
    </AppContext.Provider>
  );

  const favoriteToggler = screen.getByRole('button');

  fireEvent.click(favoriteToggler);

  expect(removeFromFavorites).toBeCalled();
});
