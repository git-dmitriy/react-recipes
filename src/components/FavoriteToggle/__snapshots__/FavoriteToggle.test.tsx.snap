// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component without errors 1`] = `
<div
  className="mx-2 mt-2"
>
  <button
    className="text-2xl text-gray-900 dark:text-orange-200 hover:text-red-500 transition duration-200 uppercase"
    onClick={[Function]}
  >
    <svg
      className="fill-current"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2 15.5V2a2 2 0 0 1 2-2h8a2 2 0 0 1 2 2v13.5a.5.5 0 0 1-.74.439L8 13.069l-5.26 2.87A.5.5 0 0 1 2 15.5zm6.5-11a.5.5 0 0 0-1 0V6H6a.5.5 0 0 0 0 1h1.5v1.5a.5.5 0 0 0 1 0V7H10a.5.5 0 0 0 0-1H8.5V4.5z"
        fillRule="evenodd"
      />
    </svg>
  </button>
</div>
`;
